webapp/src/js/services/db-sync.js

```diff
@@ -107,20 +108,114 @@ angular
       local.sync(remote);
     };
 
-    return function(updateListener) {
-      if (Session.isOnlineOnly()) {
-        if (updateListener) {
-          updateListener({ disabled: true });
-        }
+    // inProgressSync prevents multiple concurrent replications
+    var inProgressSync;
+    var sync = function(force) {
+      if (!knownOnlineState && !force) {
         return $q.resolve();
       }
 
-      syncMeta();
-      $interval(syncMeta, META_SYNC_FREQUENCY);
+      /*
+      Controllers need the status of each directed replication (directed_replication_status) and the 
+      status of the replication as a whole (aggregate_replication_status).
+      */
+      if (!inProgressSync) {
+        inProgressSync = $q
+          .all([
+            sendUpdateForDirectedReplication(replicate('from'), 'from'),
+            sendUpdateForDirectedReplication(replicateTo(), 'to'),
+          ])
+          .then(function(results) {
+            var errors = _.filter(results, function(result) {
+              return result;
+            });
+            if (errors.length > 0) {
+              $log.error('Error replicating remote server', errors);
+              sendUpdate({
+                aggregate_replication_status: 'required',
+                error: errors,
+              });
+              return;
+            }
+
+            syncIsRecent = true;
+            sendUpdate({ aggregate_replication_status: 'not_required' });
+          })
+          .finally(function() {
+            inProgressSync = undefined;
+          });
+      }
+
+      sendUpdate({ aggregate_replication_status: 'in_progress' });
+      return inProgressSync;
+    };
+
+    var sendUpdate = function(update) {
+      _.forEach(updateListeners, function(listener) {
+        listener(update);
+      });
+    };
+
+    var resetSyncInterval = function() {
+      if (intervalPromises.sync) {
+        $interval.cancel(intervalPromises.sync);
+        intervalPromises.sync = undefined;
+      }
+
+      intervalPromises.sync = $interval(function() {
+        syncIsRecent = false;
+        sync();
+      }, SYNC_INTERVAL);
+    };
+
+    return {
+      /**
+       * Adds a listener function to be notified of replication state changes.
+       *
+       * @param listener {Function} A callback `function (update)`
+       */
+      addUpdateListener: function(listener) {
+        updateListeners.push(listener);
+      },
```

Comment:
I'm wondering if it makes more sense to emit custom angular events here instead of a custom event array. It doesn't matter much currently because we're only using it in one place but it means you don't have to worry about deregistration if we listen for changes somewhere else in the app.

Reply:
Could also refactor this to a simpler `setCallback()` interface to reduce complexity (using a single function pointer instead of an array of them). May avoid complexity and the priority for deregister.

Reply:
Yeah I thought about that too but then if we _do_ use it somewhere else it wouldn't be obvious that you're essentially deregistering the original listener.

It's fine for now - we can refactor it if/when we use it somewhere else.